podutils/env.go

```diff
@@ -0,0 +1,610 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/sets"
+	apivalidation "k8s.io/apimachinery/pkg/util/validation"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/kubernetes/pkg/api/v1/resource"
+	podshelper "k8s.io/kubernetes/pkg/apis/core/pods"
+	v1helper "k8s.io/kubernetes/pkg/apis/core/v1/helper"
+	"k8s.io/kubernetes/pkg/fieldpath"
+	"k8s.io/kubernetes/pkg/kubelet/envvars"
+	"k8s.io/kubernetes/third_party/forked/golang/expansion"
+)
+
+const (
+	// ReasonOptionalConfigMapNotFound is the reason used in events emitted when an optional configmap is not found.
+	ReasonOptionalConfigMapNotFound = "OptionalConfigMapNotFound"
+	// ReasonOptionalConfigMapKeyNotFound is the reason used in events emitted when an optional configmap key is not found.
+	ReasonOptionalConfigMapKeyNotFound = "OptionalConfigMapKeyNotFound"
+
+	// ReasonOptionalSecretNotFound is the reason used in events emitted when an optional secret is not found.
+	ReasonOptionalSecretNotFound = "OptionalSecretNotFound"
+	// ReasonOptionalSecretKeyNotFound is the reason used in events emitted when an optional secret key is not found.
+	ReasonOptionalSecretKeyNotFound = "OptionalSecretKeyNotFound"
+
+	// ReasonMandatoryConfigMapNotFound is the reason used in events emitted when a mandatory configmap is not found.
+	ReasonMandatoryConfigMapNotFound = "MandatoryConfigMapNotFound"
+	// ReasonMandatoryConfigMapKeyNotFound is the reason used in events emitted when a mandatory configmap key is not found.
+	ReasonMandatoryConfigMapKeyNotFound = "MandatoryConfigMapKeyNotFound"
+	// ReasonFailedToReadMandatoryConfigMap is the reason used in events emitted when a mandatory configmap could not be read.
+	ReasonFailedToReadMandatoryConfigMap = "FailedToReadMandatoryConfigMap"
+
+	// ReasonMandatorySecretNotFound is the reason used in events emitted when a mandatory secret is not found.
+	ReasonMandatorySecretNotFound = "MandatorySecretNotFound"
+	// ReasonMandatorySecretKeyNotFound is the reason used in events emitted when a mandatory secret key is not found.
+	ReasonMandatorySecretKeyNotFound = "MandatorySecretKeyNotFound"
+	// ReasonFailedToReadMandatorySecret is the reason used in events emitted when a mandatory secret could not be read.
+	ReasonFailedToReadMandatorySecret = "FailedToReadMandatorySecret"
+
+	// ReasonInvalidEnvironmentVariableNames is the reason used in events emitted when a configmap/secret referenced in a ".spec.containers[*].envFrom" field contains invalid environment variable names.
+	ReasonInvalidEnvironmentVariableNames = "InvalidEnvironmentVariableNames"
+)
+
+var masterServices = sets.NewString("kubernetes")
+
+type ResolverConfig struct {
+	ConfigMapLister corev1listers.ConfigMapLister
+	SecretLister    corev1listers.SecretLister
+	ServiceLister   corev1listers.ServiceLister
+}
+
+func (rc ResolverConfig) CanResolveEnvVars() bool {
```

Comment:
Needs doc string

Reply:
This has been moved into the `PodController`.